motors/tacho_motor_class.c

```diff
@@ -265,6 +261,132 @@ inline struct tacho_motor_device *to_tacho_motor(struct device *dev)
+	 * decrease the speed, and use the corresponding ramp_*_sp
+	 */
+
+	if (0 <= (tm->ramp_start_speed * tm->ramp_delta_speed))
+		ramp_sp = msecs_to_jiffies(tm->active_params.ramp_up_sp);
+	else
+		ramp_sp = msecs_to_jiffies(tm->active_params.ramp_down_sp);
+
+	tm->ramp_delta_time = ramp_sp * abs(tm->ramp_delta_speed) / tm->ramp_max_speed;
+
+	/* Set the start time to about half a RAMP_PERIOD in the past so
+	 * that the first expiry starts the motor running at a non-zero
+	 * speed
+	 */
+
+	now = jiffies - RAMP_PERIOD/2;
+	tm->ramp_end_time = now + tm->ramp_delta_time;
+	tm->last_ramp_work_time = now;
+	schedule_delayed_work(&tm->ramp_work, 0);
+}
+
+static void tacho_motor_class_ramp_work(struct work_struct *work)
+{
+	struct tacho_motor_device *tm =
+		container_of(to_delayed_work(work), struct tacho_motor_device, ramp_work);
+	unsigned long remaining_ramp_time = 0;
+	unsigned long last_ramp_time;
+	int  err;
+
+	if (!IS_RUN_CMD(tm->command))
+	{
+		tm->ramp_last_speed = 0;
+		return;
+	}
+
+	/* Check to see if we are running and done ramping, or if we're
+	 * running and have hit the 0 crossover point and need to restart
+	 * the ramp
+	 */
+
+	if (tm->active_params.speed_sp == tm->ramp_last_speed) {
+		return;
+	} else if (tm->ramp_end_speed == tm->ramp_last_speed) {
+		tacho_motor_class_start_motor_ramp(tm);
+		return;
+	}
+
+	/*
+	 * If we haven't reached the end of the ramp yet, set the speed
+	 * to the appropriate point along the ramp, otherwise set the speed
+	 * directly to the setpoint. Note that we need to take into account
+	 * the fact that the target setpoint may cross the 0 point.
+	 */
+
+	if (time_is_after_jiffies(tm->ramp_end_time))
+		remaining_ramp_time = tm->ramp_end_time - jiffies;
+	else
+		remaining_ramp_time = 0;
+
+	if (tm->ramp_delta_time != 0) {
+		tm->ramp_last_speed = tm->ramp_end_speed
+			- ((tm->ramp_delta_speed * (int)remaining_ramp_time)
+				/ tm->ramp_delta_time);
+	} else {
+		tm->ramp_last_speed = tm->ramp_end_speed;
+	}
+
+	if (tm->ramp_last_speed > tm->ramp_max_speed)
+		tm->ramp_last_speed = tm->ramp_max_speed;
+	else if (tm->ramp_last_speed < -tm->ramp_max_speed)
+		tm->ramp_last_speed = -tm->ramp_max_speed;
+
+	err = tm->ops->set_speed(tm->context, tm->ramp_last_speed);
+	WARN_ONCE(err, "Failed to set speed.");
+
+	/*
+	 * Measure how long it took since the last call to ramp_work and
+	 * schedule the next ramp as close to RAMP_PERIOD as we can get
+	 */
+	last_ramp_time = jiffies - tm->last_ramp_work_time;
+	tm->last_ramp_work_time = jiffies;
+	schedule_delayed_work(&tm->ramp_work,
+		last_ramp_time >= RAMP_PERIOD ? 0 : RAMP_PERIOD - last_ramp_time);
+}
+
+static int direct_set_duty_cycle( struct tacho_motor_device *tm )
+{
+	int err = 0;
+
+	if (tm->active_params.polarity == DC_MOTOR_POLARITY_INVERSED)
+		tm->active_params.duty_cycle_sp = tm->params.duty_cycle_sp * -1;
+	else
+		tm->active_params.duty_cycle_sp = tm->params.duty_cycle_sp;
```

Comment:
This might be easier to read as 

```
tm->active_params.duty_cycle_sp = tm->params.duty_cycle_sp;
if (tm->active_params.polarity == DC_MOTOR_POLARITY_INVERSED)
	tm->params.duty_cycle_sp *= -1;
```